src/chainbin/nnet3-chain-combine.cc

```diff
@@ -19,7 +20,56 @@
 
 #include "base/kaldi-common.h"
 #include "util/common-utils.h"
-#include "nnet3/nnet-chain-combine.h"
+#include "nnet3/nnet-utils.h"
+#include "nnet3/nnet-compute.h"
+#include "nnet3/nnet-chain-diagnostics.h"
+
+
+namespace kaldi {
+namespace nnet3 {
+
+double ComputeObjf(const std::vector<NnetChainExample> &egs,
+                   NnetChainComputeProb *prob_computer) {
+  prob_computer->Reset();
+  std::vector<NnetChainExample>::const_iterator iter = egs.begin(),
+                                                 end = egs.end();
+  for (; iter != end; ++iter)
+    prob_computer->Compute(*iter);
+  const ChainObjectiveInfo *objf_info =
+      prob_computer->GetObjective("output");
+  if (objf_info == NULL)
+    KALDI_ERR << "Error getting objective info (unsuitable egs?)";
+  KALDI_ASSERT(objf_info->tot_weight > 0.0);
+  // we prefer to deal with normalized objective functions.
+  return (objf_info->tot_like + objf_info->tot_l2_term) / objf_info->tot_weight;
+}
+
+// Note: the object that prob_computer.nnet_ refers to should be
+// *moving_average_nnet.
+double UpdateNnetMovingAverageAndComputeObjf(int32 num_models,
+    const std::vector<NnetChainExample> &egs,
+    const Nnet &nnet, Nnet *moving_average_nnet,
+    NnetChainComputeProb *prob_computer) {
+  int32 num_params = NumParameters(nnet);
+  KALDI_ASSERT(num_params == NumParameters(*moving_average_nnet));
+  Vector<BaseFloat> nnet_params(num_params, kUndefined),
+      moving_average_nnet_params(num_params, kUndefined);
+  VectorizeNnet(nnet, &nnet_params);
```

Comment:
You don't need to call VectorizeNnet().  I only did that in the regular combination code to save GPU memory.  You can use the ScaleNnet and AddNnet functions declared in nnet-utils.h.